### DISCUSSION TITLE
How to match any whitespace?

### DISCUSSION THREAD

**Discussion by tisonkun** [ast-grep/ast-grep#1198] [Q&A] [answered]
I'd like to match two params of a function call, like `assert_snapshot(test("my_test_string"), @"my_expected_string");`.

But the whitespace may not be a ` ` barely, it can effectively like:

```rust
assert_snapshot!(
    test(r#"'\n'"#),
    @r###"

:'\n'
"###
);
```

and I'd rewrite it to:

```rust
assert_snapshot!(
    test(r#"'\n'"#),
    c ^ r###"

:'\n'
"###
);
```

... what pattern should I wrote?

**Comment by HerringtonDarkholme** [comment, marked as answer]
I'm assuming you are using Rust.
using [YAML rule](https://ast-grep.github.io/guide/rule-config.html), [relational rule](https://ast-grep.github.io/guide/rule-config/relational-rule.html),  and [fix region expansion](https://ast-grep.github.io/reference/yaml/fix.html#fixconfig) can solve this problem.

<img width="807" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/eabc90f8-bddf-4401-bb63-71a1d8f3fc9c">

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6Ii8vIERlYnVnLmFzc2VydCIsInJld3JpdGUiOiJEZWJ1Zy5hc3NlcnQiLCJjb25maWciOiJydWxlOlxuICBraW5kOiByYXdfc3RyaW5nX2xpdGVyYWxcbiAgcGF0dGVybjogJEFcbiAgZm9sbG93czpcbiAgICByZWdleDogJ0AnXG5maXg6XG4gIHRlbXBsYXRlOiBjIF4gJEFcbiAgZXhwYW5kU3RhcnQ6XG4gICAgcmVnZXg6ICdAJyIsInNvdXJjZSI6ImFzc2VydF9zbmFwc2hvdCEoXG4gICAgdGVzdChyI1wiJ1xcbidcIiMpLFxuICAgIEByIyMjXCJcblxuOidcXG4nXG5cIiMjI1xuKTsifQ==

**Reply by tisonkun** [reply]
Thank you! This looks reasonable. But what I need is not a "fix" but a temporary alter. That is, I need to change it back to `@r###` once some automations (namely, `cargo fmt`) done.

So here come two questions:

1. This seems replacing all `@[string-literal]`, is it possible to qualified only for the second param of `assert_snapshot`?
2. How can I write the rule in code; perhaps with `ast_grep_core`? Using CLI and config file is not quite convenient in my use case.

**Reply by HerringtonDarkholme** [reply]
For the first question, you can use [inside](https://ast-grep.github.io/reference/rule.html#inside) rule, [playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6Ii8vIERlYnVnLmFzc2VydCIsInJld3JpdGUiOiJEZWJ1Zy5hc3NlcnQiLCJjb25maWciOiJydWxlOlxuICBraW5kOiByYXdfc3RyaW5nX2xpdGVyYWxcbiAgcGF0dGVybjogJEFcbiAgZm9sbG93czpcbiAgICByZWdleDogJ0AnXG4gIGluc2lkZTpcbiAgICBwYXR0ZXJuOiBhc3NlcnRfc25hcHNob3QhKCQkJEEpXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDpcbiAgdGVtcGxhdGU6IGMgXiAkQVxuICBleHBhbmRTdGFydDpcbiAgICByZWdleDogJ0AnIiwic291cmNlIjoiYXNzZXJ0X3NuYXBzaG90IShcbiAgICB0ZXN0KHIjXCInXFxuJ1wiIyksXG4gICAgQHIjIyNcIlxuXG46J1xcbidcblwiIyMjXG4pOyJ9)

<img width="361" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/ce8dada3-bc54-405b-82a8-cf745d1f3257">

for the second, I recommend using either [JS](https://ast-grep.github.io/guide/api-usage/js-api.html) or [Py](https://ast-grep.github.io/guide/api-usage/py-api.html) API.

**Reply by tisonkun** [reply]
Thanks!

My xtask code is written by Rust and I found this API:


```rust
    let root = Rust.ast_grep("...").root();
    let forth_pattern = Op::every(KindMatcher::new("raw_string_literal", Rust))
        .and(Pattern::new("$A", Rust))
        .and(RegexMatcher::try_new(r#"@"#).unwrap());
    root.replace_all(
        forth_pattern,
        TemplateFix::try_new("c ^ $A", &Rust).unwrap(),
    );
```

It still lack:

1. How to describe `follows` and `inside` rules?
2. How to describe `expandStart` rules?
3. How to make `Vec<Edit<D>>` to string content?

**Reply by HerringtonDarkholme** [reply]
It is pretty challenging to directly use the `ast_grep_core` crate. Instead I will recommend using `ast_grep_cli` and invoke command from xtask. 

ast_grep_core has no stable API at the moment and the author does not promise any compatibility. Proceed with caution.

**Reply by tisonkun** [reply]
Good to know. Thank you!

**Comment by tisonkun** [comment]
I'm trying to change back the string literal but it failed to match:

<img width="1433" alt="image" src="https://github.com/ast-grep/ast-grep/assets/18818196/a5451e35-f48c-4037-85ee-6b72a8b5bbb9">

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6Ii8vIERlYnVnLmFzc2VydCIsInJld3JpdGUiOiJEZWJ1Zy5hc3NlcnQiLCJjb25maWciOiJpZDogYXN0LWdyZXAtYmFja1xubGFuZ3VhZ2U6IFJ1c3RcblxucnVsZTpcbiAga2luZDogcmF3X3N0cmluZ19saXRlcmFsXG4gIHBhdHRlcm46ICRBXG4gIGZvbGxvd3M6XG4gICAgcmVnZXg6ICfDpyBeICdcbiAgaW5zaWRlOlxuICAgIGFueTpcbiAgICAgIC0gcGF0dGVybjogYXNzZXJ0X3NuYXBzaG90ISgkJCRBKVxuICAgICAgLSBwYXR0ZXJuOiBhc3NlcnRfZGVidWdfc25hcHNob3QhKCQkJEEpXG4gICAgc3RvcEJ5OiBlbmRcblxuZml4OlxuICB0ZW1wbGF0ZTogQCRBXG4gIGV4cGFuZFN0YXJ0OlxuICAgIHJlZ2V4OiAnw6cgXiAnXG4iLCJzb3VyY2UiOiIjW3Rlc3RdXG5mbiB0ZXN0X3Jvd3NldCgpIHtcbiAgICBhc3NlcnRfc25hcHNob3QhKFxuICAgICAgICBSb3dTZXQ6OnRyeV9mcm9tKHZlYyFbXG4gICAgICAgICAgICB2ZWMhWzEuaW50bygpLCBcImNcIi5pbnRvKCldLmludG8oKSxcbiAgICAgICAgICAgIHZlYyFbMi5pbnRvKCksIFwiZlwiLmludG8oKV0uaW50bygpLFxuICAgICAgICAgICAgdmVjIVszLmludG8oKSwgXCJpXCIuaW50bygpXS5pbnRvKCksXG4gICAgICAgIF0pXG4gICAgICAgIC51bndyYXAoKSxcbiAgICAgICAgw6cgXiByIyMjXCJcbiAgICAgICAgICAgICstLS0tLS0tLS0tLS0rLS0tLS0tLS0tLS0tLS0tK1xuICAgICAgICAgICAgfCBjb2xfMDo6aW50IHwgY29sXzE6OnN0cmluZyB8XG4gICAgICAgICAgICArLS0tLS0tLS0tLS0tKy0tLS0tLS0tLS0tLS0tLStcbiAgICAgICAgICAgIHwgMSAgICAgICAgICB8IGMgICAgICAgICAgICAgfFxuICAgICAgICAgICAgKy0tLS0tLS0tLS0tLSstLS0tLS0tLS0tLS0tLS0rXG4gICAgICAgICAgICB8IDIgICAgICAgICAgfCBmICAgICAgICAgICAgIHxcbiAgICAgICAgICAgICstLS0tLS0tLS0tLS0rLS0tLS0tLS0tLS0tLS0tK1xuICAgICAgICAgICAgfCAzICAgICAgICAgIHwgaSAgICAgICAgICAgICB8XG4gICAgICAgICAgICArLS0tLS0tLS0tLS0tKy0tLS0tLS0tLS0tLS0tLStcbiAgICAgICAgICAgIFwiIyMjXG4gICAgKTtcbn0ifQ==

How can I match the items here?

**Reply by HerringtonDarkholme** [reply]
You can look at the playground AST dump 

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6Ii8vIERlYnVnLmFzc2VydCIsInJld3JpdGUiOiJEZWJ1Zy5hc3NlcnQiLCJjb25maWciOiJpZDogYXN0LWdyZXAtYmFja1xubGFuZ3VhZ2U6IFJ1c3RcbnJ1bGU6XG4gIGtpbmQ6IHJhd19zdHJpbmdfbGl0ZXJhbFxuICBwYXR0ZXJuOiAkQVxuICBmb2xsb3dzOlxuICAgIHJlZ2V4OiAnXidcbiAgaW5zaWRlOlxuICAgIGFueTpcbiAgICAgIC0gcGF0dGVybjogYXNzZXJ0X3NuYXBzaG90ISgkJCRBKVxuICAgIHN0b3BCeTogZW5kXG5maXg6XG4gIHRlbXBsYXRlOiAnQCRBJ1xuICBleHBhbmRTdGFydDpcbiAgICByZWdleDogw6dcbiAgICBzdG9wQnk6IGVuZFxuIiwic291cmNlIjoiI1t0ZXN0XVxuZm4gdGVzdF9yb3dzZXQoKSB7XG4gICAgYXNzZXJ0X3NuYXBzaG90IShcbiAgICAgICAgUm93U2V0Ojp0cnlfZnJvbSh2ZWMhW1xuICAgICAgICAgICAgdmVjIVsxLmludG8oKSwgXCJjXCIuaW50bygpXS5pbnRvKCksXG4gICAgICAgICAgICB2ZWMhWzIuaW50bygpLCBcImZcIi5pbnRvKCldLmludG8oKSxcbiAgICAgICAgICAgIHZlYyFbMy5pbnRvKCksIFwiaVwiLmludG8oKV0uaW50bygpLFxuICAgICAgICBdKVxuICAgICAgICAudW53cmFwKCksXG4gICAgICAgIMOnIF4gciMjI1wiXG4gICAgICAgICAgICArLS0tLS0tLS0tLS0tKy0tLS0tLS0tLS0tLS0tLStcbiAgICAgICAgICAgIHwgY29sXzA6OmludCB8IGNvbF8xOjpzdHJpbmcgfFxuICAgICAgICAgICAgKy0tLS0tLS0tLS0tLSstLS0tLS0tLS0tLS0tLS0rXG4gICAgICAgICAgICB8IDEgICAgICAgICAgfCBjICAgICAgICAgICAgIHxcbiAgICAgICAgICAgICstLS0tLS0tLS0tLS0rLS0tLS0tLS0tLS0tLS0tK1xuICAgICAgICAgICAgfCAyICAgICAgICAgIHwgZiAgICAgICAgICAgICB8XG4gICAgICAgICAgICArLS0tLS0tLS0tLS0tKy0tLS0tLS0tLS0tLS0tLStcbiAgICAgICAgICAgIHwgMyAgICAgICAgICB8IGkgICAgICAgICAgICAgfFxuICAgICAgICAgICAgKy0tLS0tLS0tLS0tLSstLS0tLS0tLS0tLS0tLS0rXG4gICAgICAgICAgICBcIiMjI1xuICAgICk7XG59In0=

**Reply by HerringtonDarkholme** [reply]
read the AST output in the bottom panel
<img width="566" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/f18264de-5e19-454f-9c4b-bb744a92a411">

and there is also https://ast-grep.github.io/reference/playground  manual

**Comment by tisonkun** [comment]
Another issue is that ast-grep accidentially format the code so that it introduced unexpected indent in raw_string_literal which changes the semantic:


```
543 543│         assert_eq!(candidates.len(), 1);
544 544│         assert_debug_snapshot!(
545 545│             candidates,
546    │-            @r###"
547    │-        [
    546│+            ç ^ r###"
    547│+                    [
548 548│             1,
549    │-        ]
550    │-        "###
    549│+                    ]
    550│+                    "###
551 551│         );
```
